### PR TITLE
Add Plugin to Extension migration guide

### PR DESCRIPTION
**Pull request by dbwiddis**
Signed-off-by: Daniel Widdis <widdis@gmail.com>

### Description

Documents the current changes needed for migration of Anomaly Detector plugin to extension

### Issues Resolved

Part of #353 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: PLUGIN_MIGRATION.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+A goal of the Extensions SDK is to ease developer migration of plugins to extensions.  Supporting this goal:
+ - Plugin interfaces and their extension points are being migrated to equivalent Extension interfaces
+ - Same-named methods and arguments are used whenever possible, with differences highlighted in this migration guide
+ - Wrapper classes are provided when possible to retain existing code logic
+
+Migration of the [Anomaly Detection Plugin](https://github.com/opensearch-project/anomaly-detection) to an Extension is in progress to both identify and refine migration challenges and provide examples to other plugin developers.
+
+## Implement Extension Interfaces and Extension Points
+
+Change the implementing class `FooPlugin` to `FooExtension` and implement the `Extension` interface.
```
Comment (ryanbogan) [inline PLUGIN_MIGRATION.md:14]: Dash before this line item?

File: PLUGIN_MIGRATION.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ - Implement other corresponding Extension interfaces, for example `ActionPlugin` interface would be `ActionExtension`.
+ - Change extension point implementation, if necessary, to conform to new types. For example, `RestHandler` classes become `ExtensionRestHandler` classes.  Consider extending base classes such as `BaseExtensionRestHandler` to provide additional convenience methods.
+
+## Use Wrapper Classes
+
+### Replace ClusterService with SDKClusterService
+
+ - `clusterService.getClusterSettings().addSettingsUpdateConsumer()` works as is, but has a method taking a map parameter can do multiple updates more efficiently.
```
Comment (ryanbogan) [inline PLUGIN_MIGRATION.md:22]: The wording of the second half of this line is a little confusing

File: PLUGIN_MIGRATION.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ - `clusterService.state()` has no changes.
+
+### Replace Client with SDKJavaClient or SDKRestClient
+
+The `SDKClient` provides two client options.
+
+The `OpenSearchClient` will be supported, and is actively developed along with other language clients and should be used whenever possible.  This client does have significant implementation differences compared to existing `Client` API.
+
+The `SDKRestClient` provides wrapper methods matching the `Client` API (but not implementing it), implemented internally with the `RestHighLevelClient`.  While this speeds migration efforts, it should be considered a temporary "bridge" with followup migration efforts to the `SDKJavaClient` planned.
+ - While the class names amd method parameters are the same, the `Request` and `Resopnse` classes are often in different packages.  In most cases, other than changing `import` statements, no additional code changes are required.  In a few changes, there are minor changes required to interface with the new response class API.
```
Comment (ryanbogan) [inline PLUGIN_MIGRATION.md:32]: I think it should be "In a few cases" instead of "In a few changes"

**Review by ryanbogan — COMMENTED** [review]
Few minor suggestions, but otherwise LGTM!

File: PLUGIN_MIGRATION.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ - `clusterService.getClusterSettings().addSettingsUpdateConsumer()` works as is, but has a method taking a map parameter can do multiple updates more efficiently.
+ - `clusterService.state()` has no changes.
+
+### Replace Client with SDKJavaClient or SDKRestClient
+
+The `SDKClient` provides two client options.
+
+The `OpenSearchClient` will be supported, and is actively developed along with other language clients and should be used whenever possible.  This client does have significant implementation differences compared to existing `Client` API.
+
+The `SDKRestClient` provides wrapper methods matching the `Client` API (but not implementing it), implemented internally with the `RestHighLevelClient`.  While this speeds migration efforts, it should be considered a temporary "bridge" with followup migration efforts to the `SDKJavaClient` planned.
```
Comment (minalsha) [inline PLUGIN_MIGRATION.md:31]: A bit confused here and please ignore my naivety. 

**OpenSearchClient**  is mentioned on line 29. In line 31, you mention SDKJavaClient. Are they 2 different clients? In Line 29, did you intend to refer SDKJavaClient?

File: PLUGIN_MIGRATION.md (hunk 2 above)
Comment (minalsha) [reply PLUGIN_MIGRATION.md:22]: +1

File: PLUGIN_MIGRATION.md (hunk 4 above)
Comment (dbwiddis) [reply PLUGIN_MIGRATION.md:31]: > Are they 2 different clients? In Line 29, did you intend to refer SDKJavaClient?

Nope, my mistake. 

There are two clients but there is no `SDKJavaClient`.  `OpenSearchClient` (the `opensearch-java` client) is the main one being used.  The method to initialize it is `SDKClient.initializeJavaClient()` and I confused myself.  It is currently a wrapper around a `RestClient` but I believe the clients team is changing that.

**Review by ryanbogan — APPROVED** [review]

File: PLUGIN_MIGRATION.md (hunk 4 above)
Comment (minalsha) [reply PLUGIN_MIGRATION.md]: This helps. Thanks

**Review by minalsha — APPROVED** [review]
LGTM

**Pull request merged**